pkg/cmd/provisioning/azure/create_managed_identities.go

```diff
@@ -0,0 +1,467 @@
+package azure
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/authorization/armauthorization"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/msi/armmsi"
+
+	credreqv1 "github.com/openshift/cloud-credential-operator/pkg/apis/cloudcredential/v1"
+	azureclients "github.com/openshift/cloud-credential-operator/pkg/azure"
+	"github.com/openshift/cloud-credential-operator/pkg/cmd/provisioning"
+
+	uuid "github.com/gofrs/uuid"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+var (
+	// CreateManagedIdentitiesOpts are azureOptions necessary for creating user-assigned managed identities
+	CreateManagedIdentitiesOpts = azureOptions{}
+
+	secretManifestTemplate = `apiVersion: v1
+stringData:
+  azure_client_id: %s
+  azure_tenant_id: %s
+  azure_region: %s
+  azure_subscription_id: %s
+  azure_federated_token_file: %s
+kind: Secret
+metadata:
+  name: %s
+  namespace: %s
+type: Opaque`
+)
+
+// createManagedIdentity creates a user-assigned managed identity for the provided CredentialsRequest
+// with name "<name>-<CredentialsRequest.Spec.SecretRef.Namespace>-<CredentialsRequest.Spec.SecretRef.Name>",
+// eg "mycluster-openshift-machine-api-azure-cloud-credentials".
+//
+// The user-assigned managed identity will be assigned pre-existing Azure roles as specified within
+// CredentialsRequest.Spec.ProviderSpec.RoleBindings. Role assignment will be scoped within the resource
+// groups provided as scopingResourceGroupNames.
+//
+// A federated identity credential will be created for each service account enumerated within
+// CredentialsRequest.Spec.ServiceAccountNames.
+//
+// A secret containing user-assigned managed identity details will be written to the outputDir
+// once the user-assigned managed identity is created and configured.
+func createManagedIdentity(client *azureclients.AzureClientWrapper, name, resourceGroupName, subscriptionID, region, issuerURL, outputDir string, scopingResourceGroupNames []string, resourceTags map[string]string, credentialsRequest *credreqv1.CredentialsRequest, dryRun bool) error {
```

Comment:
Code review: For each function where we're consuming a subset of the `azureOptions` struct, just pass the struct in.

Reply:
Also use a single validation function which does the defaulting such as deriving the OIDC resource group name from the name parameter.

Reply:
Also considering this for FUP if possible.